### ISSUE TITLE
vcpkg/scripts/detect_compiler does not exist but it does.

### ISSUE THREAD

**Issue by Daniil-Sen** [microsoft/vcpkg#17188]
When I attempt to use vcpkg install <library>, I get a cmake error in vcpkg/scripts/ports.cmake
telling me that directory vcpkg/scripts/detect_compiler does not exist. It however does exist in the scripts directory.

Here is the log content at vcpkg\buildtrees\detect_compiler\stdout-x64-windows.log :

CMake Error at C:/src/vcpkg/scripts/ports.cmake:51 (message):
  Cannot find port:

    Directory does not exist: C;/src/vcpkg/scripts/detect_compiler

My guess is for some reason a semicolon was used instead of a colon. 

- OS: Windows 10
- Compiler: MSVC
- CMake version 3.20.0

**Comment by autoantwort** [comment]
Do you only have called `./vcpkg install` or did you passed some parameters?

**Comment by dg0yt** [comment]
Really a  `;` in `C;/src/vcpkg/scripts/detect_compiler`? And not from a user-defined environment variable?

**Comment by Daniil-Sen** [comment]
> 
> 
> Really a `;` in `C;/src/vcpkg/scripts/detect_compiler`? And not from a user-defined environment variable?

Yes. I double checked my variable and it was correct. I even removed it temporarily, and tried running the vcpkg.exe with the full path from C drive in cmd. Same issue.

**Comment by Daniil-Sen** [comment]
> 
> 
> Do you only have called `./vcpkg install` or did you passed some parameters?

Just 'vcpkg install libsodium' .

**Comment by ultrasound1372** [comment]
I was getting the same thing when trying to install portaudio, #16448. I've had this bug for a few months now and no update has fixed it. So @NancyLi1013 I'm not the only one who was getting this error.

**Comment by ultrasound1372** [comment]
@Daniil-Sen Are you by chance using the cmake that comes with msys2? I looked at my debug log and found that it was the msys2 cmake that was giving that error, perhaps there's a bug in that version of cmake that doesn't like colons in its paths, even though it's supposed to support Windows.

**Comment by Daniil-Sen** [comment]
> 
> 
> @Daniil-Sen Are you by chance using the cmake that comes with msys2? I looked at my debug log and found that it was the msys2 cmake that was giving that error, perhaps there's a bug in that version of cmake that doesn't like colons in its paths, even though it's supposed to support Windows.

I'm using the one I downloaded with Cygwin.

**Comment by Daniil-Sen** [comment]
I checked which cmake I was using and it is the one I got from Cygwin.
Then I ran "vcpkg install --debug" to get some more info as to what could be happening. 

Log:

"[DEBUG] CreateProcessW(cmake -DCURRENT_PORT_DIR=C:/src/vcpkg/scripts/detect_compiler -DCURRENT_BUILDTREES_DIR=C:/src/vcpkg/buildtrees/detect_compiler -DCURRENT_PACKAGES_DIR=C:/src/vcpkg/packages/detect_compiler_x64-windows -D_HOST_TRIPLET=x64-windows -DCMD=BUILD -DDOWNLOADS=C:/src/vcpkg/downloads -DTARGET_TRIPLET=x64-windows "-DTARGET_TRIPLET_FILE=C:\src\vcpkg\triplets\x64-windows.cmake" -DVCPKG_BASE_VERSION=2021-01-13 -DVCPKG_CONCURRENCY=13 -DVCPKG_PLATFORM_TOOLSET=v142 -DVCPKG_ROOT_DIR=C:/src/vcpkg -DPACKAGES_DIR=C:/src/vcpkg/packages -DBUILDTREES_DIR=C:/src/vcpkg/buildtrees -D_VCPKG_INSTALLED_DIR=C:/src/vcpkg/installed -DDOWNLOADS=C:/src/vcpkg/downloads -DVCPKG_MANIFEST_INSTALL=OFF -P "C:\src\vcpkg\scripts\ports.cmake")
[DEBUG] CMake Error: Could not find CMAKE_ROOT !!!
[DEBUG] CMake has most likely not been installed correctly.
[DEBUG] Modules directory not found in
[DEBUG] /cygdrive/c/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20.0
[DEBUG] CMake Error at C:/src/vcpkg/scripts/ports.cmake:51 (message):
[DEBUG]   Cannot find port:
[DEBUG]
[DEBUG]     Directory does not exist: C;/src/vcpkg/scripts/detect_compiler
[DEBUG]
[DEBUG]
[DEBUG] cmd_execute_and_stream_data() returned 1 after    60195 us
[DEBUG]
[DEBUG] Compiler information tracking can be disabled by passing --feature-flags=-compilertracking"

In "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\share" I had cmake-3.19. Yet it was trying to find 3.20. That is the version from my cygwin bin.

When I changed my path for cmake from the cygwin bin directory to the one located in Microsoft Visual Studio. I ran "bootstrap-vcpkg.bat" and "vcpkg install" again. Then it worked. It looks like something is going when using cmake thats not
provided by visual studio.

**Comment by dg0yt** [comment]
> [DEBUG] CreateProcessW(cmake ...

This doesn't tell us which cmake is used. We would need to know the PATH environment variable.

I can't say much about cygwin, but for msys, it is important to use cmake from a mingw subsystem. I don't even know if cygwin provides somethin like that.

**Comment by NancyLi1013** [comment]
Hi @Daniil-Sen 

Thanks for posting this issue.

Since it's duplicated to #16448, let's track this issue there. 

Also thanks for everyone joining the discussion.

**Comment by SmileJayjhx** [comment]
I encountered this problem today, 
and the reason I suppose is my path contains Chinese characters.
I changed the Chinese words into English,
problem solved.

**Comment by binary-pixels** [comment]
string(REPLACE ";" ":" normalized_path "${CURRENT_PORT_DIR}")
    if(NOT EXISTS "${normalized_path}")
        message(FATAL_ERROR "Cannot find port: ${PORT}\n  Directory does not exist: ${normalized_path}")
    endif()
